staging/src/k8s.io/client-go/tools/portforward/portforward.go

```diff
@@ -407,11 +422,30 @@ func (pf *PortForwarder) handleConnection(conn net.Conn, port ForwardedPort) {
 	case <-localError:
 	}
 
+	// reset dataStream to discard any unsent data, preventing port forwarding from being blocked.
+	// we must reset dataStream before waiting on errorChan, otherwise,
+	// the blocking data will affect errorStream and cause <-errorChan to block indefinitely.
+	_ = dataStream.Reset()
+
 	// always expect something on errorChan (it may be nil)
 	err = <-errorChan
 	if err != nil {
 		runtime.HandleError(err)
-		pf.streamConn.Close()
+		// don't tear down the whole parent port-forward pf.streamConn when there's
+		// an error handling a single request, container runtime should handle
+		// the streaming error by reseting the connection, so we should be able
+		// to keep the connection open until a user explicitly requests an end
+		if networkClosedRe.MatchString(err.Error()) {
+			// there are two cases when we consider closing the entire connection:
+			// 1. networkClosedRe is happening when a pod is removed, in which case we should stop
+			//    port forwarding, although we'll only know about it only during subsequent
+			//    connection attempts, where one of the io.Copy goroutines fails.
+			// 2. streamTimeoutRe is happening when we're dealing with an older CRI, which doesn't
+			//    reset the streams, and thus we only learn about the error when trying to create
+			//    a subsequent stream, which fails with that error.
+			_ = pf.streamConn.Close()
```

Comment:
If you comment this out, do all tests still pass? We *have* to be sure `pf.streamConn.Close()` above as the backstop is completely sufficient and doesn't hang in new ways we don't currently hang. The only way I know of to prove that is to run all your test scenarios on master and on the PR with this commented out, and make sure we don't hang.

Once we have confidence in the backstop, then we can uncomment this and eagerly tear down earlier for specific recognized errors.